service_config.md

```diff
@@ -0,0 +1,194 @@
++-------------------------+                 +----------------------------------+
+| apiVersion: v1          |                           |
+| kind: Secret            |                           |
+| metadata:               |                           |
+|   name: service-secret2 |                           |
+|   namespace: openstack  |        1. Mounted as Volume by Service Pod
+| stringData:             |
+|   snippet3: |           |        2. `kolla_set_configs && kolla_start`: sync
+|  [formatters]           |           the resulting secret as done for the other
+|  keys=normal            |           regular config files
++-------------------------+
+```
+
+As a result of this strategy, the service presents a layout similar to the
+following:
+
+```
+00-default.conf            => default configs generated by operator. This is stored in ConfigMap
+01-deployment-secrets.conf => default configs generated by operator, which contains credentials such as [database] connection . This is stored in Secret
+02-global.conf             => custom configs provided by users via top-level customServiceConfig. This is stored in ConfigMap
+03-service.conf            => custom configs provided by users via service level customServiceConfig. This is stored in ConfigMap
+04-secrets.conf            => custom configs provided by users via service level customServiceConfigSecrets, which contains credentials. This is stored in Secret and would not be present if no secrets are provided
+```
+
+
+The service will pass the `--config-dir` parameter to point to the
+`<service>.conf.d` directory where all the config files listed above are
+rendered.
+
+```
+"command": "/usr/bin/<service> --config-dir /etc/<service>/<service>.conf.d"
+```
+
+If this strategy is not available, the `init container` that executes a start
+script (e.g., `init.sh`) won't be removed, and the logic that generates the
+layout mentioned above will be implemented in the init script.
+
+An example of the `customServiceConfigSecret` usage can be found in Manila,
+where this parameter has been used to test the
+[NetApp backend](https://gist.github.com/gouthampacha/1b5681104ee066b5bd2c702b29376199)
+
+
+## Conclusion
+
+The model described here allows to reach many goals:
+- remove (entirely) the `initContainer` and the related scripts that are no
+  longer required to start the service: **01-deployment-secret** is generated
+  by the operator according to the parameters defined in the service CR and the
+  data retrieved by the initial `osp_secret`;
+
+- `Kolla` is still used  to copy files from `src` to `dst`, as they are
+  rendered and mounted accordingly with the right permissions in the
+  destination directory (*)
+
+- Operators’ controllers are able to parse many secrets referenced by the
+  `CustomServiceConfigSecrets` parameter and merge them into a **single**
+  `Secret` which is passed to the deployment and mounted to the `Pod` in the
+  target directory (**)
```

Comment:
I'm not convinced we need to support refining multiple secrets via
CustomServiceConfigSecrets

this seems like an overcomplication to me.

what is the motivation for this functionality instead of just allowing a single secret to be used?

Reply:
`CustomServiceConfigSecrets` is an array of secrets referenced by name in the CR. 
They are merged (see my comment above) and result in a single secret mounted to the pod, but no one prevent you to use a single entry of the array.
Is this a limitation for you?

```golang
for _, secret := range secrets {
  // merge all the content coming from the referenced secret
  data := secret.Data
  for _, v := range data {
     mergedData = append(mergedData, v...)
  }
  // Define the new secret
  s := corev1.Secret{
  ObjectMeta: metav1.ObjectMeta{
	Name:      secretName,
	Namespace: instance.Namespace,
  },
  Data: map[string][]byte{
	secretName + ".conf": mergedData,
   },
}
```

Reply:
how are they mounted int eh pod? where are we specifying how to generate the kolla config json to copy them to there final location and set the permissions?

Reply:
yeah, the basic idea is to keep kolla doing the copy of the resulting (and predictable) optional file called `04-secrets.conf` (or whatever will be the name).
This means that in `pkg/<service>/volume.go` we can think about expanding `Volumes` and `VolumeMounts` arrays and mount the secrets we have (for secrets <- plural I mean `01-deployment.conf` and `04-secrets.conf` (regarding the naming, given it's internal, I'm also ok to keep something like `01-glance.conf`, `02-glance.conf`, `03-glance.conf`, `04-glance.conf` and define them as `const` in the operator, so we have a predictable amount of things we are copying).
e.g.

```golang
func GetVolumes(
	name string,
	pvcName string,
	extraVol []glancev1.GlanceExtraVolMounts,
	svc []storage.PropagationType,
	confSecret []string,
) []corev1.Volume {
    volumes := []corev1.Volume{
    ...
    ...
    ...
    }
    if len(confSecret) > 0 {
        for _, secret := range confSecret {
            sec, _ := GetConfigSecretVolumes(secret)
            volumes = append(volumes, sec)
         }
     }
    return volumes
}
``` 

where `GetConfigSecretVolumes` is a generic function that returns (corev1.Volume, corev1.VolumeMounts) associated to a generic secrets.
By doing this I can pass all the secrets generated by the operator and get them mounted to the Pod.
Kolla should be able to do the rest (it's mostly useful to adjust permissions and copy files to the `--config-dir` destination, given that mounting multiple (heterogeneous) sources in the same mountpoint doesn't work.

Reply:
i see i still dont see a stong usecase for allowing this to be an array but yes i was expecting CustomServiceConfigSecret to be the name of a single secret with a key
```
kind:secret
stringdata:
    configData: |
      [netapp]
      user=net
      password=app

```

Reply:
I think this has been covered in another thread in this patch, where @Akrog pointed out the requirement for `cinder` of  having something like:

```
cinder-volume1:

    driver-config1-secret
    fc-zone-manager-secret

cinder-volume2:

    driver-config2-secret
    fc-zone-manager-secret
``` 

(driver credentials as well as shared FC zoning driver credentials).

Reply:
I think we can resolve this comment as long as it's clear other projects (e.g. cinder) have the needs to provide an array of secrets. @SeanMooney feel free to reopen in case you have additional comments, I'm just trying to clean a bit the patch and see if we have outstanding things to cover/address